### ISSUE TITLE
Geo where clause is not working with multiple filter's

### ISSUE THREAD

**Issue by pponugo** [strongloop/loopback-datasource-juggler#931]
In the example below, when I add rental car name to search criteria, it is expected to give 'Avis' rental cars with in 5 miles of given geo location. However, find function is returning all 'Avis' rental car locations with in database. If I just pass geo location, the search is working perfectly and only returning locations with in 5 mile radius. 

I tried 'memory' as well as 'mongo', both have same issue

```
RentalLocation.find({
  // find locations near the provided GeoPoint
  where: {and:[{name:'Avis Rent A Car'}, {geo: {near: here, maxDistance: max}}]},
  // paging
  skip: limit * page,
  limit: limit
}, fn);
```

**Comment by 0candy** [comment]
Could you please fork the loopback-sandbox with a sample app to reproduce this issue? Thanks!
https://github.com/strongloop/loopback-sandbox

**Comment by etsuo** [comment]
Hi, I just discovered this ticket when working through the problem over on the loopback-connector-mongodb project: https://github.com/strongloop/loopback-connector-mongodb/issues/271

I believe I have identified the source of the problem to be that the [`nearFilter`](https://github.com/strongloop/loopback-datasource-juggler/blob/27c6279d6ce84d624457658d843967edbfd4ba61/lib/geo.js#L12) in geo.js is only checking top level properties (i.e., it's not looping recursively through the `where` object)... so it encounters an `and` and nothing else, thus reporting to the [`dao.js:find`](https://github.com/strongloop/loopback-datasource-juggler/blob/799d803fa0d6e45ec7ba3e4e000633b42945ffef/lib/dao.js#L1613) method that this is not a `near` filter, so the `connector.buildNearFilter` function of the connector is never called.

Is that the problem? Thx.

**Comment by etsuo** [comment]
@0candy, I have pushed a commit onto my fork of the `loopback-sandbox` project here: https://github.com/etsuo/loopback-sandbox

When you run the project, you should see a series of outputs:
```
$node .
Ensuring indexes exist...
done ensuring indexes.
Starting test...
Web server listening at: http://0.0.0.0:3000
Browse your REST API at http://0.0.0.0:3000/explorer
Test location created {
  "location": {
    "lat": 41.128611,
    "lng": -73.707778
  },
  "name": "test",
  "id": "57a0e0e521e8cb09a346d3e2"
}
Performing simple geo query:
Query: {
  "where": {
    "location": {
      "near": {
        "lat": 41.128611,
        "lng": -73.707778
      }
    }
  }
}
Entry found 'where near' as expected:
[
  {
    "location": {
      "lat": 41.128611,
      "lng": -73.707778
    },
    "name": "test",
    "id": "57a0e0e521e8cb09a346d3e2"
  }
]
Performing compund geo query:
Query: {
  "where": {
    "and": [
      {
        "location": {
          "near": {
            "lat": 41.128611,
            "lng": -73.707778
          }
        }
      },
      {
        "name": "test"
      }
    ]
  }
}
Expected, but undesired, error trying to performing compound geo query:
err: MongoError: error processing query: ns=test.TestTree: $and
    name == "test"
    GEONEAR  field=location maxdist=1.79769e+308 isNearSphere=0
Sort: { _id: 1 }
Proj: {}
 planner returned error: unable to find index for $geoNear query
```

If you set mongodb for profiling right before running this with `db.setProfilingLevel(2)`, you should see three entries, the latter two being the interesting ones (truncated):
```
/* 2 */
{
    "op" : "query",
    "ns" : "test.Test",
    "query" : {
        "find" : "Test",
        "filter" : {
            "location" : {
                "$near" : {
                    "$geometry" : {
                        "coordinates" : [ 
                            -73.707778, 
                            41.128611
                        ],
                        "type" : "Point"
                    }
                }
            }
        }
    }, // ... truncated 
}

/* 3 */
{
    "op" : "query",
    "ns" : "test.Test",
    "query" : {
        "find" : "Test",
        "filter" : {
            "$and" : [ 
                {
                    "location" : {
                        "$near" : {
                            "lat" : 41.128611,
                            "lng" : -73.707778
                        }
                    }
                }, 
                {
                    "name" : "test"
                }
            ]
        },
        "sort" : {
            "_id" : 1
        }
    }, // ... truncated 
```
As mentioned in my prior comment, this is because nearFilter is only checking top level properties of the where object... so it inherently doesn't recognize compound filters, so the GeoPoint isn't being replaced by GeoJSON... but this sort of filter is common, so it really should be supported.

I have a couple PRs I'm working on that fix the problem (my team's blocked on this for a project we need to deliver to a client, so we had to get some kind of workaround going)... 

How do I go about getting a set of eyes on my changes for feedback before submitting a PR? The changes will require a PR to this project as well as loopback-connector-mongodb... I need to look at the other connectors since I think my patch may have a broader impact that might not be acceptable. The way that the `nearFilter` function currently works in `geo.js`, the assumption is that there's a single key... in my mods, I changed this to an array of keys so that the `buildNearFilter` in the MongoDB connector can "walk" to the right place in the where clause to replace the GeoPoint with GeoJSON.

**Comment by 0candy** [comment]
@etsuo Thanks for creating the sample sandbox. I was able to verify the problem. Looks like in `geo.js` we only look for a simple expression or the term `near` for filter and ignore everything else ie. `and/or`.

Please submit the PR if you would like to have someone reviews the changes and give feedback. Thanks!

**Comment by etsuo** [comment]
@0candy opened PR #1028 to address this issue - I'll also opened the corresponding PR in the mongo project shortly.

**Comment by etsuo** [comment]
Opened new PR #1036

**Comment by F3L1X79** [comment]
@etsuo Any chance for this update to be finally merged since it looks like it's quite ready?

**Comment by etsuo** [comment]
Thanks @F3L1X79 - I haven't had time to review the comments made by the Loopback team. I believe there are things they want improved, which I'm happy to do (some of the stuff commented on was actually from the prior committer). 

I just need to find the time :) My travel and project schedule should settle down soon. @0candy, if I free up time to work on this, can you give me some idea of how close you are to accepting the PR for merge? If it's pretty close - I'll try to carve out some time... if it's pretty far off, then unfortunately, it'll have to wait until later. Thx.

**Comment by 0candy** [comment]
@etsuo I think you're getting there. Just need to fix up the comments from @Amir-61 and @bajtos .

**Comment by F3L1X79** [comment]
For other people's information, I tried to implement @etsuo 's fix in my own project because requests take too long if you cannot combine geolocation and other filters.
With "loopback-connector-mongodb": "^1.17.0", and "loopback-datasource-juggler": "^2.53.0" - this fix is not working, sadly.
Seems also related to https://github.com/strongloop/loopback-connector-mongodb/issues/49 but I wasn't able to anything working.

**Comment by KingsDevelopment** [comment]
Any updates on this?

**Comment by vvatelot** [comment]
Stuck with this issue too. 
Is there a workaround? How could we do to help on this subject?

**Comment by bajtos** [comment]
> How could we do to help on this subject?

@vvatelot You can take over #1036 (it looks abandoned), address code review comments and send a new pull request.

**Comment by F3L1X79** [comment]
It sounds like https://github.com/strongloop/loopback-datasource-juggler/pull/993 address the same issue.

**Comment by vvatelot** [comment]
Yes, #993 address the same issue. 
There are 2 opened PR: #993 and #1036...

Not sure what to do with this...?

**Comment by rptally** [comment]
@etsuo I just ran into this issue on my projects.  It looks like it's stuck on #1036 pending a test case and an update to this PR.  

Are you no longer pursuing these set of patches? (asks someone who really needs you to :-) )

**Comment by etsuo** [comment]
Hi @rptally - sorry, I still haven't been able to spend any time on this... work travel, projects, trying to have a personal life... etc. etc... same stuff that most trying to contribute are having to juggle. :)

Rather than using the explicit $and operator, I fell back on just relying on the implicit AND behavior of a standard query... there are circumstances when you want to use an explicit $and, but in my case it turned out that I didn't actually need it. 

Also PR https://github.com/strongloop/loopback-connector-mongodb/pull/259 was merged in and that took care of the `maxDistance` issue I was having.

If you're in a similar situation as I was in, try using the implicit AND behavior by simply not using an AND array... something like:

```
   "where": {
        "location": {
          "near": {
            "lat": 41.128611,
            "lng": -73.707778
          }
        },
        "name": "test"
    }
```

Instead of:
```
  "where": {
    "and": [
      {
        "location": {
          "near": {
            "lat": 41.128611,
            "lng": -73.707778
          }
        }
      },
      {
        "name": "test"
      }
    ]
  }
```

See: https://docs.mongodb.com/manual/reference/operator/query/and/

Hope that helps.

**Comment by rptally** [comment]
Thanks @etsuo for the quick reply… 

I’d tried the approach you recommended and couldn’t get it to work as needed. Your approach of the implied AND had two issues… one is I needed an OR and the other was it only returned 1 row / document.  After reading through this, I upgraded my loopback 2.X and related components to current, as follows:

    "loopback": "^2.36.0",
    "loopback-datasource-juggler": "^2.53.0",
    "loopback-connector-mongodb": "^1.17.0",

and tried this:

// search input object for reference
var search = {
    "term" : “kroger”,
    “location" : {
        "lat": 32.816024,
        "lng": -96.770074
    },
    "skip" : 0,
    "limit" : 10
};


// find all … near search.location for search.term in subject and body (case insensitive)
var pattern = search.term;

Collection.find({
    where : {
        and: [
            {
                updatedLoc : {
                    near: search.location
                }
            },
            {
                or : [
                    { subject:  { like: pattern, options: 'i' } },
                    { body:     { like: pattern, options: 'i' } }
                ]
            }
        ]
    },
    include : [
        'owner'
    ],
    order : [
        'updatedAt DESC'
    ],
    skip  : search.skip,
    limit : search.limit
}, 

which actually worked well and as expected.  

One additional note… Since this is for MongoDB, I had to create a “2d" index for the updatedLoc : { "type": "geopoint"  } field being used by the geo-search on the database server collection, as follows:

    db.Collection.ensureIndex({updatedLoc:"2d"});


The moral of the story is to upgrade your packages, try again, read the error messages and try again before crying for help.  :-)

Thank you!
---

**Comment by bugdanov** [comment]
Hi ! 

Same issue here, and cannot use near with an implicit "and" neither, eg
```
    where: { 
        near: { ... },
        something: true
     }
```  

I cannot get it working with @etsuo patches, and trying to use mongodb connector enableGeoIndexing options looks problematic too after reading a few issues and trying the solutions proposed (rebuilding index, using geojson) 

Will also check @zakuapps patch for this old related (duplicate?) issue: https://github.com/strongloop/loopback-connector-mongodb/issues/49

**Comment by ebarault** [comment]
hi, could you check if these PR fix your issue ?
strongloop/loopback-datasource-juggler#1216
https://github.com/strongloop/loopback-connector-mongodb/pull/322
